protocol/binary.go

```diff
@@ -61,3 +89,116 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
 	e, err := reader.ReadEnveloped()
 	return e, err
 }
+
+// DecodeRequest specializes Decode and replaces DecodeEnveloped for the
+// specific purpose of decoding request structs that may or may not have an
+// envelope.
+// This allows a Thrift request handler to transparently accept requests
+// regardless of whether the caller submits an envelope.
+// The caller specifies the expected envelope type, one of OneWay or Unary, on
+// which the decoder asserts if the envelope is present.
+func (b binaryProtocol) DecodeRequest(et wire.EnvelopeType, r io.ReaderAt) (wire.Value, EnvelopeSpecificResponder, error) {
+	// Strip request envelopes if present.
+	// 1. A message of length 1 containing only 0x00 can only be a bare,
+	// un-enveloped empty struct.
+	// 2. A message of length >1 starting with 0x00 can only be enveloped
```

Comment:
I think the explanation should include the actual expected shapes of the
payloads.

```
Enveloped (strict):
    version:2 0x00 envType:1 name~4 id:4
    where version = 0x8001
Enveloped (non-strict):
    name~4 envType:1 id:4
Non-enveloped:
    fieldType:1 (fieldID:2 fieldData:? fieldType:1)*
    when fieldType = 0, no more information follows
```

Plus it can be more obvious.

```
A message can be,

1.  Enveloped (strict)
    
    The message must begin with 0x80 -- the first byte of the version number.

2.  Enveloped (non-strict)

    The message begins with the length of the method name. As long as the
    method name is not 16 MB long, the first byte of this message is 0x00. If
    the message contains at least 9 other bytes, it is enveloped using the
    non-strict format.

        4  bytes     message length (minimum = 0)
        1  byte      message type
        4  bytes     sequence ID
        1  byte      empty struct (0x00)
        --------
        10 bytes
        
3.  Unenveloped

    If the message begins with 0x00 but does not contain any more bytes, it's
    a bare, un-enveloped empty struct.

    If the message begins with any other byte, it's an unenveloped message or
    an invalid request.
```